### PR TITLE
 BUG: non-nano strftime returns wrong results

### LINKED ISSUE #50787: BUG: non-nano strftime returns wrong results
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch] (https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
In [2]: DatetimeIndex(np.array(['-130000-01-01', '-130001-01-01'], dtype='datetime64[s]')).strftime('%Y')
Out[2]: Index(['1972', '1970'], dtype='object')
```

### Issue Description

wrong dates :)

### Expected Behavior

Index(['130000', '130001'], dtype='object')

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 9012a9f49cf34b234babb3d166bc4a8fd4f8da62
python           : 3.11.1.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.102.1-microsoft-standard-WSL2
Version          : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 2.0.0.dev0+1209.g9012a9f49c
numpy            : 1.25.0.dev0+363.gbb2769e12
pytz             : 2022.7
dateutil         : 2.8.2
setuptools       : 65.7.0
pip              : 22.3.1
Cython           : 0.29.33
pytest           : 7.2.0
hypothesis       : 6.52.1
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 8.8.0
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : None
qtpy             : None
pyqt5            : None
None

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #50787 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
@jbrockmendel I don't think there's any prospect of getting this working before 2.0.0, or is there? OK with just raising an informative message? better than giving the wrong result imo 😄

**Comment by jbrockmendel** [comment]
> better than giving the wrong result

Definitely.

It's a bit hacky, but what if we did something like

```
except ValueErrorOrWhatever:
    from pandas._libs.tslib import format_array_from_datetime
    values = np.array([self.value], dtype="i8")
    res = format_array_from_datetime(values, tz=self.tz, format=format, reso=self._creso)
    return res[0]
```

**Comment by MarcoGorelli** [comment]
thanks - just tried that and got a long error ending with
```
  File "pandas/_libs/tslibs/timestamps.pyx", line 1417, in pandas._libs.tslibs.timestamps.Timestamp.strftime
    res, = format_array_from_datetime(values, tz=self.tz, format=format, reso=self._creso)
  File "pandas/_libs/tslib.pyx", line 207, in pandas._libs.tslib.format_array_from_datetime
    res = ts.strftime(format)
  File "pandas/_libs/tslibs/timestamps.pyx", line 1417, in pandas._libs.tslibs.timestamps.Timestamp.strftime
    res, = format_array_from_datetime(values, tz=self.tz, format=format, reso=self._creso)
  File "pandas/_libs/tslib.pyx", line 207, in pandas._libs.tslib.format_array_from_datetime
    res = ts.strftime(format)
  File "pandas/_libs/tslibs/timestamps.pyx", line 1417, in pandas._libs.tslibs.timestamps.Timestamp.strftime
    res, = format_array_from_datetime(values, tz=self.tz, format=format, reso=self._creso)
  File "pandas/_libs/tslib.pyx", line 207, in pandas._libs.tslib.format_array_from_datetime
    res = ts.strftime(format)
  File "pandas/_libs/tslibs/timestamps.pyx", line 1417, in pandas._libs.tslibs.timestamps.Timestamp.strftime
    res, = format_array_from_datetime(values, tz=self.tz, format=format, reso=self._creso)
  File "pandas/_libs/tslib.pyx", line 207, in pandas._libs.tslib.format_array_from_datetime
    res = ts.strftime(format)
  File "pandas/_libs/tslibs/timestamps.pyx", line 1413, in pandas._libs.tslibs.timestamps.Timestamp.strftime
    return self.to_pydatetime().strftime(format)
  File "pandas/_libs/tslibs/timestamps.pyx", line 1095, in pandas._libs.tslibs.timestamps._Timestamp.to_pydatetime
    cpdef datetime to_pydatetime(_Timestamp self, bint warn=True):
  File "pandas/_libs/tslibs/timestamps.pyx", line 1116, in pandas._libs.tslibs.timestamps._Timestamp.to_pydatetime
    return datetime(self.year, self.month, self.day,
RecursionError: maximum recursion depth exceeded while calling a Python object
```

**Comment by jbrockmendel** [comment]
Makes sense.  Next idea would be to do something like `self.replace(year=9996).strftime(...).replace("9996", str(self.year))`

**Comment by MarcoGorelli** [comment]
maybe, but there'd be quite a lot of edge cases to take care of, and I'd be worried about rushing it

**Comment by jbrockmendel** [comment]
Agreed.

**Comment by jbrockmendel** [comment]
Maybe could handle the same subset of cases that array_strftime handles without recursing

File: pandas/_libs/tslibs/timestamps.pyx [hunk 1]
```diff
@@ -1409,7 +1409,15 @@ class Timestamp(_Timestamp):
         >>> ts.strftime('%Y-%m-%d %X')
         '2020-03-14 15:32:52'
         """
-        return datetime.strftime(self, format)
+        try:
+            return self.to_pydatetime().strftime(format)
+        except ValueError as err:
+            raise ValueError(
```
Comment (mroeschke) [inline pandas/_libs/tslibs/timestamps.pyx:1415]: `NotImplementedError`?

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/timestamps.pyx:1415]: right, that'd be better, thanks!

File: doc/source/user_guide/timeseries.rst [hunk 2]
```diff
@@ -360,6 +360,7 @@ as timezone-naive timestamps and then localize to the appropriate timezone:
    types (e.g. an int64).
 
    .. ipython:: python
+      :okwarning:
```
Comment (MarcoGorelli) [inline doc/source/user_guide/timeseries.rst:363]: This was giving
```
Out[1]: /home/marcogorelli/pandas-dev/.venv/lib/python3.8/site-packages/IPython/lib/pretty.py:778: UserWarning: Discarding nonzero nanoseconds in conversion.
  output = repr(obj)
Timestamp('2017-03-22 15:16:45.433502912')
```

Happens without IPython too:
```
>>> pd.Timestamp.min
<stdin>:1: UserWarning: Discarding nonzero nanoseconds in conversion.
Timestamp('1677-09-21 00:12:43.145224193')
```

Looking into this

**Review by mroeschke — APPROVED** [review]

**Comment by jbrockmendel** [comment]
looks like `Timestamp.__repr__` calls `strtime("%z")`.  can we support this case?

**Pull request merged**